onnxruntime/contrib_ops/cpu/ngram.cc

```diff
@@ -0,0 +1,519 @@
+ private:
+  std::vector<std::reference_wrapper<const std::string>> items_;
+  size_t hash_ = 0;
+
+  void RunningHash(const std::string& s) {
+    std::hash<std::string> hf{};
+    hash_ ^= hf(s) + 0x9e3779b9 + (hash_ << 6) + (hash_ >> 2);
+  }
+
+ public:
+  template <typename ForwardIter>
+  explicit NgramItem(size_t id, ForwardIter first, ForwardIter last) : NgramElementBase(id) {
+    while (first != last) {
+      RunningHash(*first);
+      items_.push_back(std::cref(*first));
+      ++first;
+    }
+    assert(!items_.empty());
+  }
+  explicit NgramItem() : NgramElementBase(0) {}
+  void AddItem(const std::string& s) {
+    items_.push_back(std::cref(s));
+    RunningHash(s);
+  }
+  void DebugPrint() const {
+    std::copy(items_.cbegin(), items_.cend(), std::ostream_iterator<std::string>(std::cout, ","));
+    std::cout << std::endl;
+  }
+  void Clear() {
+    items_.clear();
+    hash_ = 0;
+  }
+
+  bool operator==(const NgramItem& o) const {
+    if (items_.size() == o.items_.size()) {
+      return std::equal(items_.cbegin(), items_.cend(),
+                        o.items_.cbegin(), o.items_.cend(),
+                        std::equal_to<std::string>());
+    }
+    return false;
+  }
+  size_t Hash() const {
+    return hash_;
+  }
+};
+
+using IntegerPoolSet = std::unordered_set<NgramItem<int64_t>>;
+// Does not own strings, contains references to them. This helps
+// to search by string references that point to the current input.
+using StringPoolSet = std::unordered_set<NgramItem<std::string>>;
+
+template <typename ForwardIter, typename Cont>
+inline void Emplace(ForwardIter first, size_t ngrams, size_t ngram_size, size_t& ngram_id, Cont& c) {
+  for (; ngrams > 0; --ngrams) {
+    c.emplace(ngram_id, first, first + ngram_size);
+    first += ngram_size;
+    ++ngram_id;
+  }
+}
+
+}  // namespace ngram_details
+}  // namespace contrib
+}  // namespace onnxruntime
+
+using namespace onnxruntime::contrib::ngram_details;
+
+namespace std {
+template <typename T>
+struct hash<NgramItem<T>> {
+  typedef NgramItem<T> argument_type;
+  typedef size_t result_type;
+  result_type operator()(const argument_type& a) const {
+    return a.Hash();
+  }
+};
+}  // namespace std
+
+namespace onnxruntime {
+namespace contrib {
+
+// The weighting criteria.
+// "TF"(term frequency),
+//    the counts are propagated to output
+// "IDF"(inverse document frequency),
+//    all the counts larger than 1
+//    would be truncated to 1 and the i-th element
+//    in weights would be used to scale (by multiplication)
+//    the count of the i-th n-gram in pool
+// "TFIDF" (the combination of TF and IDF).
+//  counts are scaled by the associated values in the weights attribute.
+
+enum WeightingCriteria {
+  kNone = 0,
+  kTF = 1,
+  kIDF = 2,
+  kTFIDF = 3
+};
+
+struct Ngram::Impl {
+  WeightingCriteria weighting_criteria_ = kNone;
+  int64_t N_ = 0;
+  int64_t M_ = 0;
+  int64_t S_ = 0;
+  std::vector<int64_t> ngram_counts_;
```

Comment:
Can we at least use meaningful names here so a developer doesn't need to memorize what these stand for in order to read the code?

Reply:
This is the content of ngram_counts attribute. What is your suggestion?

---
In reply to: [243145975](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243145975) [](ancestors = 243145975)

Reply:
There are two different paths to learn a codebase. For senior people, they use code as doc so they usually start with the actual code. For junior people like me, their journeys usually start from doc and then try to connect code and its doc for better high-level understanding. I personally like the junior path because it implies a potential of having more people to contribute. If this point is also attractive to you, we can use those symbols as they are.

Reply:
Highlighting seems to have moved around. Was referring to M_, N_ and S_ not ngram_counts_. 

---
In reply to: [243365953](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243365953) [](ancestors = 243365953,243145975)

Reply:
I would prefer it if we use a meaningful name for M, N and S in the attribute name/doc and the code. In this specific case I don't see the benefit of abbreviating something like 'min' and 'max' to M and N. There are definitely times where that sort of abbreviation makes sense, but this doesn't feel like one of those to me. Is there some standard n-gram documentation where using M and N like this is the accepted norm?  

---
In reply to: [243371420](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243371420) [](ancestors = 243371420)

Reply:
Here is an example when explaining output shape of this op --- `If input shape is [B, C], this operator produces a [B, max(ngram_indexes) + 1]-tensor.` where `max` is an operator. If you change them to `max_gram_length` and `min_gram_length`, something like `Maximum number of integers to be skipped when constructing an n-gram from X. If max_skip_number=1, min_gram_length=2, max_gram_length=3, this operator may generate 2-grams with skip_number=0 and skip_number=1, and 3-grams with skip_number=0 and skip_number=1.` Is it better?

---
In reply to: [243428693](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243428693) [](ancestors = 243428693,243371420)

Reply:
I personally find that (although I'd use max_ngram_length and min_ngram_length) more readable than the following, as S, N and M have no intrinsic meaning:

"Maximum number of integers to be skipped when constructing an n-gram from X. If S=1, N=2, M=3, this operator may generate 2-grams with skip_number=0 and skip_number=1, and 3-grams with skip_number=0 and skip_number=1. "

You guys are more familiar with the usage of this so I'll leave it up to you as to whether you change it or not.

---
In reply to: [243431205](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243431205) [](ancestors = 243431205,243428693,243371420)